accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+        ChangeIt(tr);
+
+        // Case 2
+        Span<int> heap = new int[42];
+        Span<int> stack = stackalloc int[42];
+        CallMe(ref stack, heap); // okay
+        CallMe(ref heap, stack); // CS8350 as expected
+    }
+
+    public static void ChangeIt(TypedReference tr)
+    {
+        // This compiles today:
+        ref Span<int> s = ref __refvalue(tr, Span<int>);
+
+        // And can be assigned to default
+        s = default;
+
+        Span<int> newspan = stackalloc int[42];
+
+        // But this causes CS8352 as expected:
+        s = ref newspan;
+    }
+
+    public static void CallMe(ref Span<int> span1, Span<int> span2) { }
+}
+
+public ref struct MyRefStruct
+{
+    public Span<int> _span;
+}
+```
+
+If allowed today in strongly-typed manner, then reflection should also allow:
+```cs
+internal class Program
+{
+    static void Main()
+    {
+
+        Span<int> span = stackalloc int[42];
+        TypedReference tr = __makeref(span); // Assume we allow instead of CS1601
+
+        // Using a proposed invoke API; calling should be supported passing byvalue
+        MethodInfo mi1 = typeof(Program).GetMethod(nameof(ChangeIt1));
+        mi1.InvokeDirect(target: default, arg1: tr);
+
+        // and supported passing byref
+        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
+        mi2.InvokeDirect(target: default, arg1: tr);
+
+        // Just like these methods can be called today:
+        ChangeIt1(span);
+        ChangeIt2(ref span);
+    }
+
+    public static void ChangeIt1(Span<int> span) { }
+    public static void ChangeIt2(ref Span<int> span) { }
+}
+```
+
+FWIW `__arglist` currently compiles with byref-like types:
+```cs
+Span<byte> s1 = stackalloc byte[1];
+Span<byte> s2 = stackalloc byte[11];
+Span<byte> s3 = stackalloc byte[111];
+CallMe(__arglist(s1, s2, s3));
+
+static unsafe void CallMe(__arglist)
+{
+    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
+    // type although that limitation could be removed (just a runtime limitation; not compiler)
+```
+
+## Variable-length, safe collections
+(todo; see https://github.com/dotnet/runtime/issues/75349. _I have a local prototype that does a callback that doesn't require any new language or runtime features however it is not a clean programming model. The callback is necessary to interop with hide our internal GC "tracking" feature via "RegisterForGCReporting()"._)
+
+## Exception handling
+Todo; discuss using `BindingFlags.DoNotWrapExceptions` semantics only (no wrapping of exceptions).
+
+# Proposed APIs
+(work in progress; various prototypes exist here)
+## TypedReference
+```diff
+namespace System
+{
+    public ref struct TypedReference
+    {
+        // Equivalent of __makeref (except for a byref-like type since they can't be a generic parameter)
++       public static TypedReference Make<T>(ref T? value);
```

Comment:
This is strictly less powerful than `__makeref` though because it can't support `ref struct`. Until we get support for `ref struct` generic parameters my recommendation would be to use `__makeref` everywhere.

Reply:
> my recommendation would be to use __makeref everywhere.

Would we then advocate and evangelize it usage? I thought it was considered bad form to use any `__` keywords nowadays.

Reply:
One of the explicit goals of this work is to enable `ref struct` to participate here. Rather than say use `TypedReference.Make<T>` for most types but `__makeref` for `ref struct` I prefer the simpler, context-free, advice of use `__makeref`. 

I agree that `__` is explicitly unsupported now but part of this work is basically us defining these operators, adding docs and telling people when we support them.

Reply:
Based on separate discussion, my feeling here is: let's use `__makeref` etc. for the time being, but eventually make sure we introduce the ability to express `TypedReference.Make<T>` in the language in subsequent releases, and steer people towards that once it's available.